### ISSUE TITLE
Add import/export commands for backups and migration

### ISSUE THREAD

**Issue by kyrofa** [nextcloud/nextcloud-snap#185]
Should include the following:

- MySQL database
- MySQL root.ini (assuming you're dumping everything in the above step), in $SNAP_DATA/mysql
- SSL setup (private and public keys, in $SNAP_DATA/certs)
- Nextcloud config (in $SNAP_DATA/nextcloud/config)
- Nextcloud apps (in $SNAP_DATA/nextcloud/extra-apps)
- Data, optional (in $SNAP_COMMON/nextcloud)

It might make sense to make aspects of this optional.

**Comment by kapcom01** [comment]
I think this is an important feature. Right now I have to make manually a script to backup up things..

Will this command put nextcloud in maintenance mode automatically for as long as the backup process is in progress?

**Comment by kyrofa** [comment]
> Will this command put nextcloud in maintenance mode automatically for as long as the backup process is in progress?

Good question. That probably depends on what is getting fetched. If the database is included, yeah, I think that's probably the only way to get a good backup.

**Comment by KarstenDE** [comment]
Sounds very interesting. I also would prefer a "build-in" backup solution to run my environment for PROD then implement it manually.

Ideas:

- Should be able to configure it as a cron job.

- Ability to write backup to network storage. Possibly via NFS or webdav.

**Comment by kyrofa** [comment]
> Should be able to configure it as a cron job.

Oh yeah that's easy-- it'll be exported as a cli command.

> Ability to write backup to network storage. Possibly via NFS or webdav.

This is a little more difficult, as the cli commands themselves will be confined. The only way I'd be able to support being able to write anywhere is by `cat`ing the output to stdout which you could then redirect. Is that preferable over saving somewhere the snap can write and then moving it wherever you want?

**Comment by AlexanderPlaza** [comment]
In the meantime if I want to backup my data, do I just copy the data located in //snap/nextcloud/ ?
Will I need to stop the NextCloud Snap before doing that? I understand your "backing up your SD Card" documentation on the wiki, however that mostly covers if you are running it from an SD card. I am currently running it as a snap on an Ubuntu server.

**Comment by solick** [comment]
would love to have that feature. Eg. in openproject there is a cli command sudo openproject backup which creates a tarball with all necessary information. There is also a sudo openproject restore command.

**Comment by mikolodz** [comment]
@kyrofa you wrote:
>Should include the following:
MySQL database
MySQL root.ini (assuming you're dumping everything in the above step), in $SNAP_DATA/mysql
SSL setup (private and public keys, in $SNAP_DATA/certs)
Nextcloud config (in $SNAP_DATA/nextcloud/config)
Nextcloud apps (in $SNAP_DATA/nextcloud/extra-apps)
Data, optional (in $SNAP_COMMON/nextcloud)
It might make sense to make aspects of this optional.

Indeed, it says what should be backed up, but the other story is how to restore it - especially on the other machine.

I don't see it working if you move the files to the system with other nextcloud revision installed (or no revision installed). Snap knows exactly where it needs to point to with a 'current' symlink. I actually did throw all backed up folders (common & 5132) to the new destination on the second server (/var/snap/nextcloud/) and it's not even trying to work. Firstly, I did create a folder 'nextcloud' inside snap (fresh ubuntu server installation), because I hoped snap will figure it out and find imported folders, but it didn't. Then I tried to install same NextCloud revision (5132) with snap, but it is not possible as far as this version isn't even listed with 'snap info nextcloud' - so I dropped the idea. The third attempt was to create a symlink to the 5132 folder with the name 'current', but it also doesn't work. 'current' link is created by Snap while the app is being brought up and modifying it afterwards makes no sense. When the app is down, there is no 'current' symlink to be edited.

I did also try to install same revision on the newer server, but it's not possible:
>snap install nextcloud --revision 5132
error: cannot install "nextcloud": Access by specifying a revision is not allowed for this Snap.

Do you have any idea how can I force snap to recognize the existing NextCloud folder, so it tried to start it up?

**Comment by ASmith-** [comment]
Yep, how do you use the sqldump to restore the nextcloud-snap installed sql db?
My nextcloud-snap installed server has crashed 2x in the last 3 months. 

The first time it was due to a server installed faulty grub bootloader forcing me to format that servers SSD and start from scratch without the common and current folders backed up.

Then 3 days ago and 30 days rebuilding Nextcloud back and surpassing where the webserver had previously been, suddenly Nextcloud refused to recognize the valid and still current certs for the domain and lets encrypt refused to upgrade. I backed up the snap nextcloud current and common folders plus performed a sqldump as well. I once again re-formated the servers SSD and reinstalled its linux distro from scratch. After installing snapd, nextcloud and logging in using the same username, same password as before and then on the command line checking to make certain lets encrypt functioned properly with my domain by adding a current, functioning cert before moving to restore.

I have restored the current folder and the common folders using a current backup just prior to formating the SSD on the server, however Nextcloud after doing so now hangs at the login. Nextcloud appears to remember which email addresses were used to register new users but will not accept any logins. This is likely a sign the sql db file is not properly restored. "How do I use the sqldump file to restore the Nextcloud sql database file I had previously" ?

**Comment by kyrofa** [comment]
Could use some help testing #688 for anyone interested in this functionality (details for how to test on the PR).
